### PR TITLE
Add transforms' forward_shape and inverse_shape

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR steals @fritzo implementation in https://github.com/pytorch/pytorch/pull/50581 and resolve various issues that couldn't be addressed previously. Thanks for your effort, @fritzo!

### CODE REVIEW

File: numpyro/distributions/distribution.py [hunk 1]
```diff
@@ -774,16 +762,10 @@ def __init__(self, base_distribution, transforms, validate_args=None):
...
-        if event_dim <= base_ndim:
-            batch_shape = shape[:base_ndim - event_dim]
-            event_shape = shape[base_ndim - event_dim:]
-        else:
-            event_shape = (-1,) * event_dim
-            batch_shape = ()
+        shape = transform.forward_shape(shape)
+        cut = len(shape) - event_dim
+        batch_shape = shape[:cut]
+        event_shape = shape[cut:]
```
Comment (fritzo) [inline numpyro/distributions/distribution.py:768]: Nice, this is so much simpler that the PyTorch version! I think I'll copy your use of `ComposeTransform` upstream in https://github.com/pytorch/pytorch/pull/50581 😄

**Review by fritzo — APPROVED** [review]
Thanks for porting this! My only suggestion is to add a comment about `.forward_shape()` versus `.forward_event_shape()`.

File: numpyro/distributions/distribution.py (hunk 1 above)
Comment (fritzo) [reply numpyro/distributions/distribution.py:768]: @fehiepsi I've made some more changes to the upstream version, including wrapping `base_dist` in `Independent` to ensure `.sample(sample_shape)` is correct. Can you PTAL at https://github.com/pytorch/pytorch/pull/50581?

File: numpyro/contrib/tfp/distributions.py [hunk 2]
```diff
@@ -86,6 +86,12 @@ def _inverse(self, y):
     def log_abs_det_jacobian(self, x, y, intermediates=None):
         return self.bijector.forward_log_det_jacobian(x, self.domain.event_dim)
 
+    def forward_shape(self, shape):
+        return self.bijector.forward_event_shape(shape)
+
+    def inverse_shape(self, shape):
+        return self.bijector.inverse_event_shape(shape)
+
```
[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/contrib/tfp/distributions.py:94]: You are right. Bijectors like `Affine(zeros(3), ones(3))` do not align with Pyro transforms. How do you know?

I just added a fix for it.

File: numpyro/distributions/distribution.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/distributions/distribution.py:768]: You are right. We have assumed too much on base distribution. Your change is pretty clean. Thanks!

File: numpyro/contrib/tfp/distributions.py [hunk 3]
```diff
@@ -86,6 +86,18 @@ def _inverse(self, y):
     def log_abs_det_jacobian(self, x, y, intermediates=None):
         return self.bijector.forward_log_det_jacobian(x, self.domain.event_dim)
 
+    def forward_shape(self, shape):
+        out_shape = self.bijector.forward_event_shape(shape)
+        in_event_shape = self.bijector.inverse_event_shape(out_shape)
+        batch_shape = shape[:len(shape) - len(in_event_shape)]
+        return batch_shape + out_shape
```
Comment (fritzo) [inline numpyro/contrib/tfp/distributions.py:93]: Nice trick!

**Review by fritzo — APPROVED** [review]
Thanks for porting this!

**Pull request merged**